_posts/2023-03-26-Kafka-IaC.md

```diff
@@ -0,0 +1,249 @@
+    configNonsensitive: {
+      ["connector.class"]: "AzureDataLakeGen2Sink", # The class of the Connector. List of [supported connectors](https://docs.confluent.io/cloud/current/connectors/index.html#supported-connectors)
+      ["name"]: "Connector Name",
+      ["kafka.auth.mode"]: "KAFKA_API_KEY",
+      ["topics"]: topicNames,
+      ["input.data.format"]: "JSON",
+      ["output.data.format"]: "JSON",
+      ["time.interval"]: "HOURLY",
+      ["tasks.max"]: "2",
+      ["flush.size"]: "1000",
+      ["rotate.schedule.interval.ms"]: "3600000",
+      ["rotate.interval.ms"]: "3600000",
+      ["path.format"]: "'year'=YYYY/'month'=MM/'day'=dd/'hour'=HH",
+      ["topics.dir"]: "<Directory in ADLS>",
+    },
+    configSensitive: {
+      ["kafka.api.key"]: kafka_api_key,
+      ["kafka.api.secret"]: kafka_api_secret,
+      ["azure.datalake.gen2.account.name"]: azure_data_lake_account_name,
+      ["azure.datalake.gen2.access.key"]: azure_data_lake_access_key,
+    },
+    environment: cluster_environment,
+    kafkaCluster: cluster,
+  };
+
+  new confluent.Connector("pulumi-connector", connector_args);
+```
+
+If the Confluent Cloud cluster credentials are already set up, directly go ahead and run `pulumi up` command to provision the topics in the confluent cloud. If the Confluent Cloud cluster credentials aren’t set up, follow the steps from the topic provisioning and set them up.
+
+## Terraform
+
+Terraform is the leading IaC tool available in the market with support to various cloud, datacenter and services. It supports many cloud computing platforms like Azure, AWS, Oracle, Google and Container orchestration like Kubernetes etc. Read about the supported providers [here](https://registry.terraform.io/browse/providers). Infact pulumi provider is built on top of the official Confluent Terraform Provider. Infrastructure is provisioned with Terraform using HashiCorp Configuration Language (HCL). 
+
+### Provisioning Topics
+
+First, Initialize the provider with
+
+```
+terraform {
+  required_providers {
+    confluent = {
+      source  = "confluentinc/confluent"
+      version = "1.13.0"
+    }
+  }
+}
+```
+This installs the confluent cloud provider.
+
+Configure the confluent secrets with
+
+```
+provider "confluent" {
+  cloud_api_key    = var.confluent_cloud_api_key    # optionally use CONFLUENT_CLOUD_API_KEY env var
+  cloud_api_secret = var.confluent_cloud_api_secret # optionally use CONFLUENT_CLOUD_API_SECRET env var
+}
+```
+
+Once the Confluent provider and the configurations are set up, run `terraform init`
+
+```
+resource "confluent_kafka_topic" "dev_topics" {
+  kafka_cluster {
+    id = var.cluster_id
+  }
+  for_each         = toset(var.topics)
+  topic_name       = each.value
+  rest_endpoint    = data.confluent_kafka_cluster.dev_cluster.rest_endpoint
+  partitions_count = 6
+  config = {
+    "retention.ms" = "604800000"
+  }
+  credentials {
+    key    = var.api_key
+    secret = var.api_secret
+  }
+}
+```
+
```

Comment:
Conclusion on the three different IaC and suggestions on when to use them. A scenario where one of them is more appropriate than others